### PR TITLE
[web][SIMS #695] Create common component for student application

### PR DESCRIPTION
**Pull request by ann-aot**
- Created a common shared component for student Applications and it's shared between AEST and Students
- code smell due to todo (TODO is for me 😄)

### CODE REVIEW

File: sources/packages/web/src/router/StudentRoutes.ts [hunk 1]
```diff
@@ -2,11 +2,6 @@ import { RouteRecordRaw } from "vue-router";
 import StudentDashboard from "@/views/student/StudentDashboard.vue";
 import Login from "@/views/student/Login.vue";
 import AppStudent from "@/views/student/AppStudent.vue";
-import FinancialAidApplication from "@/views/student/financial-aid-application/FinancialAidApplication.vue";
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/router/StudentRoutes.ts:5]: Thanks for doing the cleanup 😉

File: sources/packages/web/src/components/common/StudentApplication.vue [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+  <formio
+    :formName="selectedForm"
+    :data="initialData"
+    :readOnly="isReadOnly"
+    @loaded="formLoaded"
+    @changed="formChanged"
+    @submitted="submitted"
+    @customEvent="customEvent"
+  ></formio>
+  <StudentApplicationCommonActions
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/common/StudentApplication.vue:11]: I still do not feel the need of creating this component. Is it going to grow or have much more logic that justifies it?

File: sources/packages/web/src/views/aest/StudentApplicationView.vue [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+      required: true,
+    },
+  },
+  setup(props: any) {
+    const applicationDetail = ref({} as GetApplicationBaseDTO);
+    const initialData = ref({});
+    const selectedForm = ref();
+
+    onMounted(async () => {
+      applicationDetail.value = await ApplicationService.shared.getApplicationDetail(
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/aest/StudentApplicationView.vue:60]: Probably not part of this PR, but why are we using the _applicationId_ and _studentId_?
From the API point of view does not make sense IMO and it is forcing the component to receive both as a property.
@ann-aot, I took a look in the API and it seems that it would be worth changing it.

File: sources/packages/web/src/components/common/StudentApplication.vue [hunk 4]
```diff
@@ -0,0 +1,273 @@
+<template>
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/common/StudentApplication.vue:1]: Should we move this component to the partials folder maybe?

**Review by andrewsignori-aot — COMMENTED** [review]
Overall looks good, a great start. Please take a look at the comments.

File: sources/packages/web/src/components/common/StudentApplication.vue [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+      required: true,
+    },
+    programYearId: {
+      type: Number,
+      required: true,
+    },
+  },
+  setup(props: any, context: SetupContext) {
+    // Components names on Form.IO definition that will be manipulated.
+    const LOCATIONS_DROPDOWN_KEY = "selectedLocation";
```
Comment (henryoforeh-aot) [inline sources/packages/web/src/components/common/StudentApplication.vue:65]: A small typo here. It should be Component's names

File: sources/packages/web/src/components/common/StudentApplication.vue [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+      type: Boolean,
+      required: true,
+    },
+    programYearId: {
+      type: Number,
+      required: true,
+    },
+  },
+  setup(props: any, context: SetupContext) {
+    // Components names on Form.IO definition that will be manipulated.
```
Comment (henryoforeh-aot) [inline sources/packages/web/src/components/common/StudentApplication.vue:64]: A small typo here. It should be Component's names

File: sources/packages/web/src/components/common/StudentApplicationCommonActions.vue [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    <v-col md="6">
+      <v-btn
+        color="primary"
+        v-show="!isFirstPage"
+        outlined
+        @click="$emit('wizardGoPrevious')"
+        >Previous section</v-btn
+      >
+    </v-col>
+    <v-col md="6" class="ml-auto text-right">
```
Comment (dheepak-aot) [inline sources/packages/web/src/components/common/StudentApplicationCommonActions.vue:12]: @ann-aot when we spoke during sync up our plan to keep these actions inside the application common component right? Is there any specific reason to keep actions as separate component?

File: sources/packages/web/src/components/common/StudentApplication.vue [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+      type: Object,
+      required: true,
+    },
+    selectedForm: {
+      type: String,
+      required: true,
+    },
+    isReadOnly: {
+      type: Boolean,
+      required: true,
```
Comment (dheepak-aot) [inline sources/packages/web/src/components/common/StudentApplication.vue:56]: isReadOnly is this required property ?

Comment (ann-aot) [reply sources/packages/web/src/components/common/StudentApplication.vue:56]: yes, the value is passes from the parent component

File: sources/packages/web/src/components/common/StudentApplication.vue [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+      };
+      applicationWizard.on("prevPage", prevNextNavigation);
+      applicationWizard.on("nextPage", prevNextNavigation);
+
+      // Emit setFirstLastPage event to the parent, so that the
+      // respective values are set on parent side
+      context.emit("setFirstLastPage", isFirstPage.value, isLastPage.value);
+
+      // TODO:Ann check below code for AEST
+      await formioDataLoader.loadLocations(form, LOCATIONS_DROPDOWN_KEY);
```
Comment (dheepak-aot) [inline sources/packages/web/src/components/common/StudentApplication.vue:117]: Do we still require to load the drop down for readOnly part?

Comment (dheepak-aot) [reply sources/packages/web/src/components/common/StudentApplication.vue:117]: To be more clear, I am talking about using the label approach.

File: sources/packages/web/src/components/common/StudentApplicationCommonActions.vue (hunk 7 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/common/StudentApplicationCommonActions.vue:12]: Yeah got that it is the child of common component itself. Just trying to understand, Is there any specific reason to keep actions as separate component?

**Comment by dheepak-aot** [comment]
Good work @ann-aot 👍Making student application as common component is a challenge which has been handled very nicely.
2 comments from me.

File: sources/packages/web/src/components/common/StudentApplication.vue [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+import { ref, SetupContext } from "vue";
+import {
+  useFormioComponentLoader,
+  useFormioDropdownLoader,
+  useFormioUtils,
+} from "@/composables";
+
+export default {
+  emits: [
+    "loadForm",
```
[the remark this answers is not in the document]
Comment (ann-aot) [reply sources/packages/web/src/components/common/StudentApplication.vue:36]: renamed to `formLoadedCallback` as the common component already have a function called `formLoaded`

File: sources/packages/web/src/components/common/StudentApplicationCommonActions.vue (hunk 7 above)
Comment (ann-aot) [reply sources/packages/web/src/components/common/StudentApplicationCommonActions.vue:12]: since in the student application all action buttons are inside vue, i wanted to keep it as a separate action component specific for the student application, that was my reason,  but updated the code

File: sources/packages/web/src/components/common/StudentApplication.vue (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/common/StudentApplication.vue:1]: As agreed, we will be talking about it in the next DEV sync-up.

**Review by andrewsignori-aot — APPROVED** [review]
looks good 👍

**Review by guru-aot — APPROVED** [review]
LGTM, Nice work @ann-aot

**Review by dheepak-aot — APPROVED** [review]
LGTM. 👍
